tests/python/unittest/test_kvstore_custom.py

```diff
@@ -0,0 +1,165 @@
+def test_pushpull_single_kv_pair():
+    """aggregate value on muliple devices"""
+    def check_aggregator(kv, key, key_list=None):
+        kv.broadcast(key, mx.nd.zeros(shape), out=mx.nd.empty(shape))
+        # devices
+        num_devs = 4
+        devs = [mx.Context('cpu', i) for i in range(num_devs)]
+
+        # single
+        vals = [mx.nd.ones(shape, d) for d in devs]
+        outs = [mx.nd.empty(shape, d) for d in devs]
+
+        kv.pushpull(key, vals, out=outs)
+        for out in outs:
+            check_diff_to_scalar(out, num_devs)
+
+        # inplace
+        kv.pushpull(key, vals)
+        for val in vals:
+            check_diff_to_scalar(val, num_devs)
+
+        # list
+        if key_list is None:
+            return
+        num_keys = len(key_list)
+        kv.broadcast(key_list, [mx.nd.zeros(shape)] * num_keys,
+                     out=[mx.nd.empty(shape)] * num_keys)
+        vals = [[mx.nd.ones(shape, d)*2.0 for d in devs]] * num_keys
+        outs = [[mx.nd.empty(shape, d) for d in devs]] * num_keys
+        kv.pushpull(key_list, vals, out=outs)
+        for out in outs:
+            for o in out:
+                check_diff_to_scalar(o, num_devs * 2.0)
+
+        # inplace
+        kv.pushpull(key_list, vals)
+        for val in vals:
+            for v in val:
+                check_diff_to_scalar(v, num_devs * 2.0)
+
+    check_aggregator(init_kv('device'), 3, keys)
+    check_aggregator(init_kv('device'), 'a', str_keys)
+    check_aggregator(init_kv('teststore'), 3)
+    check_aggregator(init_kv('teststore'), 'a')
+
+@with_seed()
+def test_pushpull_list_kv_pair():
+    """aggregate value on muliple devices"""
+    def check_aggregator(kv, key, key_list=None):
+        kv.broadcast(key, mx.nd.zeros(shape), out=mx.nd.empty(shape))
+        # devices
+        num_devs = 4
+        devs = [mx.Context('cpu', i) for i in range(num_devs)]
+
+        # single
+        vals = [mx.nd.ones(shape, d) for d in devs]
+        outs = [mx.nd.empty(shape, d) for d in devs]
+
+        kv.pushpull(key, vals, out=outs)
+        for out in outs:
+            check_diff_to_scalar(out, num_devs)
+
+        # list
+        if key_list is None:
+            return
+        num_keys = len(key_list)
+        kv.broadcast(key_list, [mx.nd.zeros(shape)] * num_keys,
+                     out=[mx.nd.empty(shape)] * num_keys)
+        vals = [[mx.nd.ones(shape, d)*2.0 for d in devs]] * num_keys
+        outs = [[mx.nd.empty(shape, d) for d in devs]] * num_keys
+        kv.pushpull(key_list, vals, out=outs)
+        for out in outs:
+            for o in out:
+                check_diff_to_scalar(o, num_devs * 2.0)
+
+    check_aggregator(init_kv('device'), 3, keys)
+    check_aggregator(init_kv('device'), 'a', str_keys)
+    check_aggregator(init_kv('teststore'), 3)
+    check_aggregator(init_kv('teststore'), 'a')
+
+@with_seed()
+def test_get_type_device():
+    kvtype = 'device'
+    kv = mx.kv.create(kvtype)
+    assert kv.type == kvtype
+
```

Comment:
Do we have to test set_optimizer_states and load_optimizer_states as well?

Reply:
i can add that